opm/utility/ECLGraph.hpp

```diff
@@ -0,0 +1,154 @@
+        ECLGraph() = delete;
+
+        /// Destructor.
+        ~ECLGraph();
+
+        /// Move constructor.
+        ///
+        /// \param[in] rhs Graph from which to appropriate resources.
+        /// Invalid upon return.
+        ECLGraph(ECLGraph&& rhs);
+
+        /// Disabled copy constructor
+        ECLGraph(const ECLGraph& rhs) = delete;
+
+        /// Move assignment operator.
+        ///
+        /// \param[in] rhs Graph from which to appropriate resources.
+        ///
+        /// \return Reference to \code *this \endcode.
+        ECLGraph& operator=(ECLGraph&& rhs);
+
+        /// Disabled assignment operator.
+        ECLGraph& operator=(const ECLGraph& rhs) = delete;
+
+        /// Named constructor.
+        ///
+        /// \param[in] grid Name or prefix of ECL grid (i.e., .GRID or
+        ///                 .EGRID) file.
+        ///
+        /// \param[in] init Name of ECL INIT file corresponding to \p grid
+        ///                 input.  Assumed to provide at least a complete
+        ///                 set of pore-volume values (i.e., for all global
+        ///                 cells defined in the \p grid).
+        ///
+        ///                 If available in the INIT file, the constructor
+        ///                 will also leverage the transmissibility data
+        ///                 when constructing the active cell neighbourship
+        ///                 table.
+        ///
+        /// \return Fully formed ECLIPSE connection graph with property
+        /// associations.
+        static ECLGraph
+        load(const Path& grid, const Path& init);
+
+        /// Assign source object for phase flux calculation.
+        ///
+        /// \param[in] src Name of ECL restart file, possibly unified, from
+        ///                which next set of phase fluxes should be retrieved.
+        void assignFluxDataSource(const Path& src);
+
+        /// Retrieve number of active cells in graph.
+        std::size_t numCells() const;
+
+        /// Retrive number of connections in graph.
+        std::size_t numConnections() const;
+
+        /// Retrive neighbourship relations between active cells.
+        ///
+        /// The \c i-th connection is between active cells \code
+        /// neighbours()[2*i + 0] \endcode and \code neighbours()[2*i + 1]
+        /// \endcode.
+        const std::vector<int>& neighbours() const;
+
+        /// Retrive static pore-volume values on active cells only.
+        ///
+        /// Corresponds to the \c PORV vector in the INIT file, possibly
+        /// restricted to those active cells for which the pore-volume is
+        /// strictly positive.
+        const std::vector<double>& poreVolume() const;
+
+        /// Retrive phase flux on all connections defined by \code
+        /// neighbours() \endcode.
+        ///
+        /// Non-"const" because this potentially loads new data from the
+        /// backing store into internal cache data structures.
```

Comment:
Not sure what holes you mean? If the caches could be moved to a separate class (they might already be, have not read the code enough yet), then just that object could be mutable and it would be very much isolated.

The disadvantage of the current solution is of course that one could accidentally change other parts of the object (other than the cache).